src/DotNetty.Codecs/LengthFieldBasedFrameDecoder.cs

```diff
@@ -0,0 +1,450 @@
+            if (input.ReadableBytes < frameLengthInt)
+                return null;
+
+            if (this._initialBytesToStrip > frameLengthInt)
+            {
+                input.SkipBytes(frameLengthInt);
+                throw new CorruptedFrameException("Adjusted frame length (" + frameLength + ") is less " +
+                    "than initialBytesToStrip: " + this._initialBytesToStrip);
+            }
+            input.SkipBytes(this._initialBytesToStrip);
+
+            // extract frame
+            int readerIndex = input.ReaderIndex;
+            int actualFrameLength = frameLengthInt - this._initialBytesToStrip;
+            IByteBuffer frame = this.ExtractFrame(context, input, readerIndex, actualFrameLength);
+            input.SetReaderIndex(readerIndex + actualFrameLength);
+            return frame;
+        }
+
+        /// <summary>
+        ///  Decodes the specified region of the buffer into an unadjusted frame length.  The default implementation is
+        /// capable of decoding the specified region into an unsigned 8/16/24/32/64 bit integer.  Override this method to
+        /// decode the length field encoded differently. 
+        /// Note that this method must not modify the state of the specified buffer (e.g. <see cref="IByteBuffer.ReaderIndex"/>, 
+        /// <see cref="IByteBuffer.WriterIndex"/>, and the content of the buffer.)
+        /// </summary>
+        /// <param name="buffer">The buffer we'll be extracting the frame length from.</param>
+        /// <param name="offset">The offset from the absolute <see cref="IByteBuffer.ReaderIndex"/>.</param>
+        /// <param name="length">The length of the framelenght field. Expected: 1, 2, 3, 4, or 8.</param>
+        /// <param name="order">The preferred <see cref="ByteOrder"/> of <see cref="buffer"/>.</param>
+        /// <returns>A long integer that represents the unadjusted length of the next frame.</returns>
+        protected long GetUnadjustedFrameLength(IByteBuffer buffer, int offset, int length, ByteOrder order)
+        {
+            buffer = buffer.WithOrder(order);
+            long frameLength;
+            switch (length)
+            {
+                case 1:
+                    frameLength = buffer.GetByte(offset);
+                    break;
+                case 2:
+                    frameLength = buffer.GetShort(offset);
+                    break;
+                case 4:
+                    frameLength = buffer.GetInt(offset);
+                    break;
+                case 8:
+                    frameLength = buffer.GetLong(offset);
+                    break;
+                default:
+                    throw new DecoderException("unsupported lengthFieldLength: " + this._lengthFieldLength + " (expected: 1, 2, 3, 4, or 8)");
+            }
+            return frameLength;
+        }
+
+        protected IByteBuffer ExtractFrame(IChannelHandlerContext context, IByteBuffer buffer, int index, int length)
```

Comment:
[nit] will become virtual to support overrides in MarshallingDecoder and ObjectDecoder